### PR TITLE
[CUDA][SDPA] Fix stride usage in mem-efficient attention

### LINKED ISSUE #191937: SDPA EFFICIENT_ATTENTION returns wrong output for broadcasted (stride-0) k/v when kv length % 32 == 1 (bf16 + attn_mask, CUDA; MATH correct on identical tensors)
### 🐛 Describe the bug

This report is the result of a long debugging session, where I was heavily involved. The AI had an earlier hypothesis which the experiments rejected, and I personally ran and reviewed every experiment described in this post. Most of this text and the code are prepared by Claude Fable through Cursor, but I have read the full text and it is correct on all points.

`F.scaled_dot_product_attention` with the **EFFICIENT_ATTENTION** backend
returns badly corrupted output (max abs error ~19.5, vs ~0.09 honest bf16
wobble) for a specific real-world input whose **kv sequence length is
≡ 1 mod 32** (bf16, CUDA, batch 1, 4-D boolean `attn_mask`). The **MATH**
backend on the *identical* tensors is correct (matches an fp32 reference
to bf16 wobble), and the *same* tensors with **one extra kv position**
(kv_len ≡ 2 mod 32) are clean on every backend. The attached repro
replays all four selectable backends, so the comparison is complete on
any hardware (CUDNN and FLASH decline these inputs: head_dim 512,
non-null mask).

**The memory layout matters as much as the values.** The failing call
(a Gemma 4 KV-shared/GQA attention layer) looks like:

```
q         bf16 shape (1, 16, 289, 512) stride (2367488, 512, 8192, 1)  # non-contiguous [B,S,H,D] permute
k, v      bf16 shape (1, 16, 289, 512) stride (0, 0, 512, 1)           # ONE kv head expand()ed to 16 heads
attn_mask bool shape (1, 1, 289, 289)  contiguous
dropout_p=0.0  scale=1.0  is_causal=False
```

Materializing the same values contiguously (a `.cpu().clone()` round
trip) makes EFFICIENT_ATTENTION **correct** — the bug requires the
broadcasted stride-0 k/v (and/or the strided q). It is also
value-dependent: randomly initialized weights did not reproduce during
transformers-side triage. The repro therefore ships the exact captured
tensors as raw storages + `(shape, stride, storage_offset)` and rebuilds
the precise views with `Tensor.set_`; this layout-exact replay reproduces
the corrupted full-model output **bit-exactly** (max abs diff 0.0 vs the
captured forward).

Found via Gemma 4 batched generation in transformers
(huggingface/transformers#47651): left-padding a prompt so the padded
total hits 1 mod 32 flips the next-token argmax to an unrelated special
token (logit deltas ~40). Maintainer triage there ruled transformers out:
eager attention and SDPA-with-MATH are clean and the presence of images
is irrelevant.

Repro (only `torch` imported; the `.pt` loads with `weights_only=True`):

```bash
unzip sdpa_repro.zip
python torch_sdpa_repro.py sdpa_repro.pt
```

The script replays both captured calls under every selectable backend
(MATH bf16, EFFICIENT, CUDNN, FLASH) against an fp32-MATH reference and
prints max abs differences. Output on the reporting machine:

```
torch==2.12.0+cu130  cuda=13.0  gpu=NVIDIA RTX PRO 6000 Blackwell Workstation Edition
bundle meta: {'source': 'https://github.com/huggingface/transformers/issues/47651', 'model': 'google/gemma-4-12B-it', 'torch': '2.12.0+cu130', 'transformers': '5.14.1', 'gpu': 'NVIDIA RTX PRO 6000 Blackwell Workstation Edition', 'call_index': 41, 'n_calls_per_prefill': 48, 'poison_kv_len': 289, 'control_kv_len': 290, 'guilty_backends': ['EFFICIENT'], 'note': 'poisoned kv_len % 32 == 1; control differs by ONE left-pad token, same layer, same prompt. Tensors are packed as raw storages + (shape, stride, offset) because the bug is layout-sensitive.'} 

[poisoned] q shape (1, 16, 289, 512) kv_len 289 (% 32 = 1) dtype bfloat16
  captured original run vs fp32 ref:    19.5110
  MATH_bf16  vs fp32 ref:     0.0625
  EFFICIENT  vs fp32 ref:    19.5110
  CUDNN      unsupported on these inputs (RuntimeError)
  FLASH      unsupported on these inputs (RuntimeError)

[control] q shape (1, 16, 290, 512) kv_len 290 (% 32 = 2) dtype bfloat16
  captured original run vs fp32 ref:     0.0865
  MATH_bf16  vs fp32 ref:     0.0625
  EFFICIENT  vs fp32 ref:     0.0865
  CUDNN      unsupported on these inputs (RuntimeError)
  FLASH      unsupported on these inputs (RuntimeError)

Expected: every backend matches the fp32-MATH reference to bf16 wobble (the MATH_bf16 row) on both cases. BUG: on the poisoned case (kv_len % 32 == 1) at least one backend diverges orders of magnitude past wobble -- matching the 'captured original run' row, which is what the full model forward produced -- while the control (ONE extra kv position, otherwise the same tensors) is clean everywhere.
```

The two calls come from the same attention layer of the same prompt and
differ only by one left-pad token, so the control isolates the kv_len
alignment as the trigger. The capture harness (how the tensors were
extracted from the model forward, plus a sweep showing the 1-mod-32
pattern across many lengths) is in the transformers issue linked above.

### Expected behavior

Every backend matches MATH to numerical wobble at every sequence length,
as they already do for the control case and for every other length in
the sweeps on the transformers side (the pattern held across prompt
lengths ~280–2900: corruption iff total kv_len % 32 == 1).

Observed on: torch 2.12.0+cu130, NVIDIA RTX PRO 6000 Blackwell
Workstation Edition, bf16, single GPU. Not yet tried on non-Blackwell
hardware (the transformers maintainer who reproduced the pattern can
confirm their hardware in the linked issue).

**AI disclosure** (per this repo's AI policy): the repro scripts and the
first draft of this report were written with AI assistance (Claude, via
Cursor). I ran every experiment myself on the hardware above — all
numbers in this report are from my own runs — and the
layout-sensitivity finding came from iterating on two earlier capture
designs whose replays failed to reproduce the corruption until the
exact storages/strides were preserved. I have reviewed the scripts and
the analysis, understand them, and take responsibility for the report.

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @drisspg @liangel-02 @howardzhang-cv (SDPA / attention)

### Versions

Collecting environment information...
PyTorch version: 2.12.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A 

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.13 | packaged by conda-forge | (main, Mar  5 2026, 16:50:00) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.17.0-35-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 13.0.88
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA RTX PRO 6000 Blackwell Workstation Edition
Nvidia driver version: 580.159.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.14.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.14.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.14.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.14.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.14.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.14.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.14.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.14.0
Is XPU available: False
HIP runtime version: N/A 
MIOpen runtime version: N/A 
Is XNNPACK available: True
Caching allocator config: N/A 

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           48 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  16  
On-line CPU(s) list:                     0-15
Vendor ID:                               AuthenticAMD
Model name:                              AMD Ryzen 7 9700X 8-Core Processor
CPU family:                              26  
Model:                                   68  
Thread(s) per core:                      2   
Core(s) per socket:                      8   
Socket(s):                               1   
Stepping:                                0   
Frequency boost:                         enabled
CPU(s) scaling MHz:                      52% 
CPU max MHz:                             5582.3008
CPU min MHz:                             605.3100
BogoMIPS:                                7585.25
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good amd_lbr_v2 nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpuid_fault cpb cat_l3 cdp_l3 hw_pstate ssbd mba perfmon_v2 ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local user_shstk avx_vnni avx512_bf16 clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif x2avic v_spec_ctrl vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid bus_lock_detect movdiri movdir64b overflow_recov succor smca fsrm avx512_vp2intersect flush_l1d amd_lbr_pmc_freeze
Virtualization:                          AMD-V
L1d cache:                               384 KiB (8 instances)
L1i cache:                               256 KiB (8 instances)
L2 cache:                                8 MiB (8 instances)
L3 cache:                                32 MiB (1 instance)
NUMA node(s):                            1   
NUMA node0 CPU(s):                       0-15
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; IBPB on VMEXIT only
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; STIBP always-on; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB on VMEXIT

Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] onnxruntime==1.28.0
[pip3] torch==2.12.0+cu130
[pip3] torchcodec==0.12.0+cu130
[pip3] torchvision==0.27.0+cu130
[pip3] triton==3.7.0
[conda] numpy                        2.4.6            pypi_0                pypi
[conda] nvidia-cublas                13.1.1.3         pypi_0                pypi
[conda] nvidia-cuda-cupti            13.0.85          pypi_0                pypi
[conda] nvidia-cuda-nvrtc            13.0.88          pypi_0                pypi
[conda] nvidia-cuda-runtime          13.0.96          pypi_0                pypi
[conda] nvidia-cudnn-cu13            9.20.0.48        pypi_0                pypi
[conda] nvidia-cufft                 12.0.0.61        pypi_0                pypi
[conda] nvidia-curand                10.4.0.35        pypi_0                pypi
[conda] nvidia-cusolver              12.0.4.66        pypi_0                pypi
[conda] nvidia-cusparse              12.6.3.3         pypi_0                pypi
[conda] nvidia-cusparselt-cu13       0.8.1            pypi_0                pypi
[conda] nvidia-nccl-cu13             2.29.7           pypi_0                pypi
[conda] nvidia-nvjitlink             13.0.88          pypi_0                pypi
[conda] nvidia-nvtx                  13.0.85          pypi_0                pypi
[conda] torch                        2.12.0+cu130     pypi_0                pypi
[conda] torchcodec                   0.12.0+cu130     pypi_0                pypi
[conda] torchvision                  0.27.0+cu130     pypi_0                pypi
[conda] triton                       3.7.0            pypi_0                pypi

[sdpa_repro.zip](https://github.com/user-attachments/files/30668089/sdpa_repro.zip)

### PR DESCRIPTION
**Pull request by eqy**
To fix #191937 , repro ultimately didn't need `.pt` or hardcoded data that the user provided

authored with codex
CC @malfet 

cc @ptrblck @msaroufim @tinglvv @nWEIdia @drisspg @liangel-02 @howardzhang-cv

### CODE REVIEW

**Review by drisspg — APPROVED** [review]
at some point we should properly setup layouts to handle all the indexing; 

Also, ideally we should get to a place where are confident enough in cudnn and PT's support matrix for hardware that we only need math, fa, and CuDNN

**Comment by drisspg** [comment]
@claude effort=low can you look for other instances or similar like mis-indexing in the efficient attention kernel, please and thanks boss

**Comment by Skylion007** [comment]
@claude can you draft the fixes and also add minimal unit tests for these edge cases?

**Comment by drisspg** [comment]
@Skylion007  I went through and addressed all and some more / verified here: https://github.com/pytorch/pytorch/pull/192138
